### ISSUE TITLE
Why we are removing the repositories after the installation ?

### ISSUE THREAD

**Issue by vhsantos** [wazuh/wazuh-ansible#1050]
Hello there,

Anyone know why we are removing the repositories (deb/rpm) after the installation ??
https://github.com/wazuh/wazuh-ansible/blob/53ccf753be00a7ac767a3a88e92d4cdb63d00a7c/roles/wazuh/ansible-wazuh-manager/tasks/main.yml#L337-L339

Maybe the idea is to not allow automatic update using apt update/upgrade, but, not make many sense for me !! :-(

Any comment ?

thanks and attentive.

**Comment by rauldpm** [comment]
Yes, the main reason is to avoid possible unwanted component upgrades when performing general upgrades, since in many cases the upgrade processes involve a procedure to maintain the post-upgrade functionality.

Looking at the incorporation date of the task (2020) you can see that at that time the Wazuh indexer/dashboard components were not installed, but rather ElasticSearch was used, in addition, the other component that was installed was the Wazuh agent which, since it cannot be installed in the same instance as the manager, it made no sense that the upgrade would affect it.

Now, with the possible installation of the Wazuh indexer/dashboard on the same instance as the Wazuh manager, an upgrade could be catastrophic if not controlled

On the other hand, I see that there is a difference between installation by sources and by package, which is why I have opened this issue so the team can work on it, as the processes should have the same ending: https://github.com/wazuh/wazuh-ansible/issues/1052

I have also detected a problem when installing from sources if the version of GCC that the instance has is not correct: https://github.com/wazuh/wazuh-ansible/issues/1051, if you are going to carry out this type of deployment please check the requirements: https://documentation.wazuh.com/current/deployment-options/wazuh-from-sources/wazuh-server/index.html#installing-dependencies

**Comment by vhsantos** [comment]
Hello there @rauldpm 

Ok.. It is like I have imagined (some manual pre/post update tasks).  :-(

I know that "sometimes" exists big changes when one new mayor/minor versions is out, but **"in theory"** security updates (5.4.XXX) should not break anything on the systems and it should be applied as soon as possible. So,  I wanted to discuss an approach that could simplify our system update process while maintaining a high level of security.

Instead of removing the repository from our systems, we consider configuring our update process to focus solely on security updates. This way, we can strike a balance between security and efficiency. Here's how it could work:

```bash
# cat /etc/apt/preferences.d/wazuh.pref 
Package: wazuh-*
Pin: origin packages.wazuh.com
Pin: version 4.4*
Pin-Priority: 1001

Package: filebeat-*
Pin: origin packages.wazuh.com
Pin: version 7.10*
Pin-Priority: 1001

# the same to others packages on the repo
# and a general rule to avoid new packages not configured before
Package: *
Pin: origin packages.wazuh.com
Pin-Priority: 1
```

With this configuration, we could enable unattended-upgrades on our systems, allowing them to automatically receive and apply **only** security updates. and/or users can also run apt update/upgrade as part of regular maintenance without needing to modify repository files.  ;-)

This apt setting is easy to implement and we can make it optional.

Attentive to any comments and ideas.

Note: I'm not sure, but I hope that Wazuh didn't have any issues or compatibility problems in the past when applying security fixes/updates that might have discouraged us from pursuing this idea. ;-)

**Comment by rauldpm** [comment]
If we were talking about the unattended upgrade of a Wazuh manager without having Filebeat, Wazuh indexer, or Wazuh dashboard deployed, there would be no problem in performing the upgrades (except if it is an agent since the version of the agent must be the same or lower than that of Wazuh manager), but please note that the MAJOR.MINOR of the Wazuh manager must be the same as the Wazuh indexer and Wazuh dashboard versions

It is not about the upgrade changes breaking the components, but about the relationship that exists between the components, that is why I have mentioned that when the said task was included, there was no such dependency with the Wazuh indexer and Wazuh dashboard since they are more recent components

Perhaps you might be better interested in a modification like the following to prevent components from being upgraded in an unwanted way.

```
sed -i "s/^deb/#deb/" /etc/apt/sources.list.d/wazuh.list
apt-get update
```

or

```
apt-mark hold wazuh-*
```

Please review the following documentation to see what an upgrade of the Wazuh components entails: https://documentation.wazuh.com/current/upgrade-guide/upgrading-central-components.html#upgrading-the-wazuh-indexer

Please note that the upgrade process varies depending on the component:

- In the case of Wazuh manager (as a unitary component) it is not necessary to perform additional actions, but if it is installed together with Filebeat it is necessary to update, for example, the alert template
- In the case of the Wazuh indexer, it is necessary and advisable to deactivate the shard allocation during the process, in addition to the fact that each cluster node must be upgraded sequentially in a way that allows minimal disruption of the service
- In the case of the Wazuh dashboard no change is required
- Regarding the Wazuh indexer and Wazuh dashboard, you must keep in mind that manually installed plugins must be upgraded manually since they are not part of the package and this means that the upgrade process may fail.


In general, an unattended upgrade is valid and recommended in certain cases, such as when the component does not have dependencies on other components, such as an installation of Wazuh manager without Wazuh indexer, Filebeat, and Wazuh dashboard, but when in the environment, there are relationships between different components, upgrades should be supervised, since in general and regardless of the changes they introduce, these relationships exist between the components.

**Comment by vhsantos** [comment]
Hello Raul,

I'm talking about **all packages** inside of the **wazuh repository (packages.wazuh.com)**, but, specifically, I'm referring to security/patch updates only, not upgrades that involve major/minor version changes. IMHO, these patches should be applied as soon as possible and should not introduce any backward compatibility issues.

Imagine a scenario where a zero-day vulnerability is discovered in one of the Wazuh packages. A security/patch update, let's say version 4.5.3, is released on a Friday at 02:00 AM, but it might not be installed until Monday morning! This delay can be concerning.


I've considered your two suggestions beofre my first message:

1. Disabling the repository: While this is an option, it would require users to uncomment the repository before being able to use commands like apt install wazuh-manager=4.5.3. This extra step may not be convenient.

2. Using apt hold: This approach is similar to disabling the repository since users would need to "unhold" all packages before performing an update.

With the preferences file, we can set all packages from this repository to be not able to update (inclusive with the `apt upgrade`). However, they would still be available when users or ansible playbooks want to install or upgrade/update it, the only difference is that will be necessary to pass the version as `apt install wazuh-manager=4.5.3` (note: ansible playbook already do it using the variable **wazuh_manager_version**):

```bash
# and a general rule to avoid new packages not configured before
Package: *
Pin: origin packages.wazuh.com
Pin-Priority: 1
```

I have also reviewed the links you provided:

[Upgrading Central Components](https://documentation.wazuh.com/current/upgrade-guide/upgrading-central-components.html)
[Compatibility Matrix](https://documentation.wazuh.com/current/upgrade-guide/compatibility-matrix/index.html)

While I acknowledge these resources, I still believe that the software should maintain compatibility within the same major.minor versions and allow for some differences between patch versions.

After considering your very welcome comments and suggestions, I am inclined to propose this change within the Wazuh project itself rather than within the Wazuh Ansible repository (because it looks more a design than a automation issue). I will make a similar suggestion to the developers for their consideration.

1000 Thanks

**Comment by rauldpm** [comment]
Hello, thanks for the issue!

Regard:

>Ideally, I would expect Wazuh software to exhibit compatibility across patch versions. For instance:
>
>Wazuh Manager 4.5.3 should work seamlessly with Wazuh Indexer 4.5.X and Wazuh Dashboard 4.5.X.
>Wazuh Manager 4.5.3 should be compatible with Wazuh Agent versions 4.5.4 or 4.5.X.

I must clarify that they should be compatible between patches, but not between minors, perhaps it should have been clarified better in this reference.

> but please note that the MAJOR.MINOR of the Wazuh manager must be the same as the Wazuh indexer and Wazuh dashboard versions

#### A Wazuh manager 4.4.5 should work with a Wazuh dashboard 4.4.3 package, but ideally, it is recommended that the versions be the same

As I mentioned, the problem is not with the versions (except when there is a minor change), but with the Wazuh indexer upgrade process and the Filebeat template.
